lib/screens/config/cache.ex

```diff
@@ -0,0 +1,180 @@
+
+  def ok?, do: table_exists?()
+
+  def refresh_if_loaded_before(screen_id) do
+    with_table default: nil do
+      [[last_deploy_timestamp]] = :ets.match(@table, {:last_deploy_timestamp, :"$1"})
+
+      refresh_if_loaded_before =
+        case :ets.match(@table, {screen_id, %{refresh_if_loaded_before: :"$1"}}) do
+          [[timestamp]] -> timestamp
+          [] -> nil
+        end
+
+      case {last_deploy_timestamp, refresh_if_loaded_before} do
+        {nil, nil} ->
+          nil
+
+        {nil, refresh_if_loaded_before} ->
+          refresh_if_loaded_before
+
+        {last_deploy_timestamp, nil} ->
+          last_deploy_timestamp
+
+        {last_deploy_timestamp, refresh_if_loaded_before} ->
+          Enum.max([last_deploy_timestamp, refresh_if_loaded_before], DateTime)
+      end
+    end
+  end
+
+  def last_deploy_timestamp do
+    with_table do
+      [[last_deploy_timestamp]] = :ets.match(@table, {:last_deploy_timestamp, :"$1"})
+      last_deploy_timestamp
+    end
+  end
+
+  def service_level(screen_id) do
+    with_table default: 1 do
+      case :ets.match(@table, {{:screen, screen_id}, %{app_params: %{service_level: :"$1"}}}) do
+        [[service_level]] -> service_level
+        [] -> 1
+      end
+    end
+  end
+
+  def disabled?(screen_id) do
+    with_table default: false do
+      case :ets.match(@table, {{:screen, screen_id}, %{disabled: :"$1"}}) do
+        [[disabled]] -> disabled
+        [] -> false
+      end
+    end
+  end
+
+  def screen(screen_id) do
+    with_table default: nil do
+      case :ets.match(@table, {{:screen, screen_id}, :"$1"}) do
+        [[screen]] -> screen
+        [] -> nil
+      end
+    end
+  end
+
+  def app_params(screen_id) do
+    with_table default: nil do
+      case :ets.match(@table, {{:screen, screen_id}, %{app_params: :"$1"}}) do
+        [[app_params]] -> app_params
+        [] -> nil
+      end
+    end
+  end
+
+  def devops do
+    with_table default: nil do
+      case :ets.match(@table, {:devops, :"$1"}) do
+        [[devops]] -> devops
+        [] -> nil
+      end
+    end
+  end
+
+  @doc """
+  Returns a list of all screen IDs.
+  """
+  def screen_ids do
+    with_table default: [] do
+      # Not sure how to avoid matching against the struct this way.
+      # Since this isn't a true pattern match, doing %Screen{} actually tries to instantiate the struct,
+      # and we end up matching against all of its default field values,
+      # or getting a compile error if it has any enforced keys.
+      @table
+      |> :ets.match({{:screen, :"$1"}, :_})
+      |> List.flatten()
+    end
+  end
+
+  @doc """
+  Returns a list of all screen IDs that satisfy the given filter.
+  The filter function will be passed a tuple of {screen_id, screen_config} and should return true if that screen ID should be included in the results.
+  """
+  def screen_ids(filter_fn) do
+    with_table default: [] do
+      filter_reducer = fn
```

Comment:
I don't think there's anything wrong with using `foldl` here, but the function I mentioned in another comment, `fun2ms`, might reduce the amount of code needed here. Unless there is a substantial performance impact, totally up to you which you use.

Reply:
I think `fun2ms` imposes the same limitations on what kinds of checks/guards you can do, since it translates `fun` to a match spec. `fun` can't call an arbitrary `filter_fn`, it can only use `is_*` guards.

From the docs:
> The fun is very restricted, it can take only a single parameter (the object to match): a sole variable or a tuple. It must use the is_ guard tests. Language constructs that have no representation in a match specification (if, case, receive, and so on) are not allowed.